server/athenian/api/experiments/preloading/cache.py

```diff
@@ -0,0 +1,259 @@
+        if total:
+            size = sum(s)
+            return size if not human else _humanize_size(size)
+        else:
+            return s if not human else s.apply(_humanize_size)
+
+    def _squeeze(self, df: pd.DataFrame) -> pd.DataFrame:
+        def _cast_category(df, cols):
+            for c in cols:
+                df[c.name] = df[c.name].astype("category")
+            return df
+
+        def _cast_timestamp(df, cols):
+            for c in cols:
+                df[c.name] = (df[c.name].astype("int64") / 10 ** 9).astype("int32")
+                df.loc[df[c.name] < 0, c.name] = 0
+            return df
+
+        def _cast_identifiers(df, cols):
+            for c in cols:
+                max_length = df[c.name].apply(len).max()
+                df[c.name] = df[c.name].astype(f"|S{max_length}")
+            return df
+
+        return (
+            df.pipe(_cast_category, self._categorical_cols)
+            .pipe(_cast_timestamp, self._timestamp_cols)
+            .pipe(_cast_identifiers, self._identifier_cols)
+        )
+
+    def __repr__(self):
+        """Return the representation of the `CachedDataFrame`."""
+
+        def fmt_cols(cols):
+            return ", ".join(c.name for c in cols)
+
+        metadata = [
+            f"Table: {self._db.url}",
+            f"Columns: {fmt_cols(self._cols)}",
+            f"Timestamp columns: {fmt_cols(self._timestamp_cols)}",
+            f"Categorical columns: {fmt_cols(self._categorical_cols)}",
+            f"Identifier columns: {fmt_cols(self._identifier_cols)}",
+        ]
+
+        lines = metadata + ["DataFrame", self._df.__repr__()]
+        return "\n".join(lines)
+
+    def __getattr__(self, attr):
+        """Make attributes access propagate to the wrapped DataFrame."""
+        if self._df is None:
+            raise ValueError("DataFrame not loaded")
+
+        return self._df.__getattr__(attr)
+
+    def __getitem__(self, key):
+        """Make items access propagate to the wrapped DataFrame."""
+        if self._df is None:
+            raise ValueError("DataFrame not loaded")
+
+        return self._df.__getitem__(key)
+
+
+class MemoryCache:
+    """A singleton in-memory cache for instances of `CachedDataFrame`."""
+
+    _instance = None
+
+    def __init__(
+        self,
+        mdb: DatabaseLike,
+        pdb: DatabaseLike,
+        options: Dict[str, Dict],
+        debug_memory: Optional[bool] = False,
+    ):
+        """Initialize a `MemoryCache`."""
+        self._mdb = mdb
+        self._pdb = pdb
+        self._options = options
+        self._debug_memory = debug_memory
+        self._dfs = {}
+
+    @classmethod
+    def get_instance(
+        cls,
+        mdb: Optional[DatabaseLike] = None,
+        pdb: Optional[DatabaseLike] = None,
+        options: Optional[Dict[str, Dict]] = None,
+        debug_memory: Optional[bool] = False,
+    ) -> "MemoryCache":
+        """Return the singleton instance."""
+        if cls._instance is None:
```

Comment:
What if we avoid the singleton pattern like this:

- Create `MemoryCache` in our `_preload_dataframes` callback.
- 
```
for db in (mdb, pdb):
    db.cache = memory_cache
```
- Change `ParallelDatabase` to redirect queries to that cache. We already intercept all the queries there, so it shouldn't be too hard.

Reply:
The first two points are trivial and I agree with them, it would be cleaner. 👍 

> We already intercept all the queries there, so it shouldn't be too hard

This is an interesting approach 🤔 .

If we already intercept the queries (sorry I didn't see that part of the code yet), it should be pretty easy to then use [pandasql](https://github.com/yhat/pandasql/) that I mentioned to you in one of our calls already: so we should be able to just compile the SQLA query and pass it to pandasql. In case of an error of missing table/dataframe then it will fall back to the "normal" DB query. I never used pandasql and I'm not much familiar with query compilation, but worth exploring this option. This should also save "translating" the current queries into pandas operations as you can see [here in this draft PR](https://github.com/athenianco/athenian-api/pull/1337/files#diff-0ad06f21667ae6b7fa99290cd35ebd84fdcab18d79d7711e6c505449bb328e9fR70).

I'm just wondering whether there are some weird cases that we might bump into by using pandasql such as unsupported things. Now that I'm checking probably it won't be that straightforward since it works with SQLite syntax, but in production, the query would be prepared for PG, so we would need to prepare the query for both dialects I guess?

For sure it's something that I'd like to have long-term: a unified interface with SQL query that is aware of whether to query the in-mem cache or the underlying DB. But I'm wondering whether this can at least wait for once we see the gain on at least the first fully-speeded-up endpoint before investing the time on this.

wdyt?

Reply:
Anyway I'm exploring this idea since it turned out that it's super simple compile the query in any dialect:
```
In [24]: from sqlalchemy.dialects import postgresql, sqlite

In [25]: from athenian.api.models.metadata.github import PullRequest

In [26]: query = sa.select([PullRequest.acc_id, PullRequest.node_id]).where(PullRequest.acc_id.in_([1,2,3]))

In [27]: str(query.compile(dialect=postgresql.dialect(), compile_kwargs={"literal_binds": True}))
Out[27]: 'SELECT github.api_pull_requests.acc_id, github.api_pull_requests.node_id \nFROM github.api_pull_requests \nWHERE github.api_pull_requests.acc_id IN (1, 2, 3)'

In [28]: str(query.compile(dialect=sqlite.dialect(), compile_kwargs={"literal_binds": True}))
Out[28]: 'SELECT github.api_pull_requests.acc_id, github.api_pull_requests.node_id \nFROM github.api_pull_requests \nWHERE github.api_pull_requests.acc_id IN (1, 2, 3)'
```

I'm gonna try this approach + `pandasql` in the draft PR to see whether a real query works as expected.

Reply:
damn I just realized that the latest release for `pandasql` was in 2016... and py2.7 only 🤦

Reply:
ok nvm the code is super tiny, there shouldn't be any issue, I'll test it. When installing it just builds the wheel without issues 👍

Reply:
`literal_binds` does not always work, unfortunately. We use some Postgres and SQLite specific extensions that do not compile in that mode, for example, HSTORE, JSON, ARRAY. AFAIR even datetimes do not inline well. This is a piece of serious work. I studied the machinery in fine detail while maintaining the Postgres backend of `databases`. The low-level asyncpg code deals with SQL string and a series of arguments. The serialization of the arguments can be stolen from asyncpg, I think. Basically, we need our own asyncpg.Connection imposter. And yes, some engine to convert either SQLAlchemy or raw SQL to pandas commands.

So you originally planned to swap DB queries with pandas without hacking the DB interface by rewriting each query manually, I see.

Reply:
Ah damn, I forgot about those extensions. As I mentioned in DM despite I'd love to have a common interface in the future, the effort that might be required for doing it properly _now_ scares me. I'd rather go easy for now and re-evaluate later on once we are benefiting already from the performance speedup.

In any case, I'll have a look at attaching the `MemoryCache` to the databases as you proposed 👍